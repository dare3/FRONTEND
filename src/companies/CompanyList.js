import React, { useState, useEffect, useRef } from 'react'; // Import useRef
import SearchBar from '../forms/SearchBar';
import JoblyApi from '../common/api';
import CompanyCard from './CompanyCard';
import LoadingSpinner from '../common/LoadingSpinner';

/** Show list of all companies.
 *
 * On mount, loads companies from API.
 * Re-loads filtered companies on submit from search form.
 *
 * Routes -> { CompanyCard, SearchForm }
 *
 * Routed to at /companies
 */

const CompanyList = () => {
  const [companies, setCompanies] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  const isMounted = useRef(true); // Create a ref to track mounted status

  useEffect(() => {
    isMounted.current = true; // Set to true when component mounts

    async function getCompaniesOnMount() {
      console.debug('CompanyList useEffect getCompaniesOnMount');
      await search();
    }

    getCompaniesOnMount();

    return () => {
      isMounted.current = false; // Cleanup function to mark as unmounted
    };
  }, []);

  /** Triggered by search form submit; reloads companies. */
  async function search(name) {
    try {
      let companies = await JoblyApi.getCompanies(name);
      if (isMounted.current) { // Check if still mounted
        setCompanies(companies);
      }
    } catch (error) {
      console.error("Failed to fetch companies:", error);
      // Optionally, set an error state here
    } finally {
      if (isMounted.current) {
        setIsLoading(false);
      }
    }
  }

  if (isLoading) {
    return <LoadingSpinner message="Loading companies..." />;
  }

  return (
    <div className="CompanyList col-md-8 offset-md-2">
      <SearchBar searchFor={search} />
      {companies.length ? (
        <div className="CompanyList-list">
          {companies.map(c => (
            <CompanyCard
              key={c.handle}
              handle={c.handle}
              name={c.name}
              description={c.description}
              logoUrl={c.logoUrl}
            />
          ))}
        </div>
      ) : (
        <p className="message">Sorry, no results were found!</p>
      )}
    </div>
  );
};

export default CompanyList;
