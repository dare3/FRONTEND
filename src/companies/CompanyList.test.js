import React from 'react';
import { render, screen } from '@testing-library/react';
import CompanyList from './CompanyList';
import JoblyApi from '../common/api';

// Mocking JoblyApi.getCompanies
jest.mock('../common/api');

describe('CompanyList', () => {
  beforeEach(() => {
    jest.clearAllMocks(); // Clear mocks before each test
  });

  test('renders without errors', () => {
    JoblyApi.getCompanies.mockResolvedValue([]); // Mock API response
    render(<CompanyList />);
  });

  test('matches snapshot', () => {
    JoblyApi.getCompanies.mockResolvedValue([]); // Mock API response
    const { asFragment } = render(<CompanyList />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('displays loading spinner initially', () => {
    JoblyApi.getCompanies.mockImplementation(() => new Promise(() => {})); // Mocking a pending promise
    render(<CompanyList />);
    expect(screen.getByText(/loading/i)).toBeInTheDocument(); // Check for loading spinner text
  });

  test('displays no results message when no companies are found', async () => {
    JoblyApi.getCompanies.mockResolvedValue([]); // Mock API response
    render(<CompanyList />);
    
    // Wait for companies to be loaded
    expect(await screen.findByText(/sorry, no results were found!/i)).toBeInTheDocument();
  });

  test('displays companies when fetched', async () => {
    const mockCompanies = [
      { handle: 'company1', name: 'Company One', description: 'Description One', logoUrl: '' },
      { handle: 'company2', name: 'Company Two', description: 'Description Two', logoUrl: '' },
    ];
    JoblyApi.getCompanies.mockResolvedValue(mockCompanies); // Mock API response

    render(<CompanyList />);
    
    // Wait for companies to be loaded
    expect(await screen.findByText(/company one/i)).toBeInTheDocument();
    expect(await screen.findByText(/company two/i)).toBeInTheDocument();
  });
});
