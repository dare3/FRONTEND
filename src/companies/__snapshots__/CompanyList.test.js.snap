// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyList matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="LoadingSpinner"
  >
    Loading companies...
  </div>
</DocumentFragment>
`;
